system_hooks/src/call_hooks/l1_messenger.rs

```diff
@@ -132,175 +128,47 @@ where
-            ))
-        }
-    };
-    if abi_encoded_message_len < length_encoding_end {
-        return Ok(Err(
-            "L1 messenger failure: sendToL1 called with invalid calldata",
-        ));
-    }
-    let length: u32 = match U256::from_be_slice(
-        &abi_encoded_message[(length_encoding_end as usize) - 32..length_encoding_end as usize],
-    )
-    .try_into()
-    {
-        Ok(length) => length,
-        Err(_) => {
-            return Ok(Err(
-                "L1 messenger failure: sendToL1 called with invalid calldata",
-            ))
-        }
-    };
-    // to check that it will not overflow
-    let message_end = match length_encoding_end.checked_add(length) {
-        Some(message_end) => message_end,
-        None => {
-            return Ok(Err(
-                "L1 messenger failure: sendToL1 called with invalid calldata",
-            ))
-        }
-    };
-    if abi_encoded_message_len < message_end {
-        return Ok(Err(
-            "L1 messenger failure: sendToL1 called with invalid calldata",
-        ));
-    }
+    let address_sender = B160::try_from_be_slice(&calldata[0..20]).ok_or(
+        SystemError::LeafDefect(internal_error!("Failed to create B160 from 20 byte array")),
+    )?;
 
-    // Note, that in general, Solidity allows to have non-strict offsets, i.e. it should be possible
-    // to call a function with offset pointing to a faraway point in calldata. However,
-    // when explicitly calling a contract Solidity encodes it via a strict encoding and allowing
-    // only standard encoding here allows for cheaper and easier implementation.
-    if abi_encoded_message_len % 32 != 0 {
-        return Ok(Err("Calldata is not well formed"));
-    }
+    let message = &calldata[20..];
 
-    let message = &abi_encoded_message[(length_encoding_end as usize)..message_end as usize];
-    // Charge gas for l1 message
-    let l1_message_cost_ergs = l1_message_ergs_cost(message.len());
-    resources.charge(&S::Resources::from_ergs(l1_message_cost_ergs))?;
+    // emit L1 message
     let message_hash = system.io.emit_l1_message(
-        // We already charged gas for it
+        // Gas should be charged by the L1Messenger system contract
         ExecutionEnvironmentType::NoEE,
         resources,
-        &caller,
+        &address_sender,
         message,
     )?;
 
-    let mut topics = ArrayVec::<Bytes32, MAX_EVENT_TOPICS>::new();
-    topics.push(Bytes32::from_array(L1_MESSAGE_SENT_TOPIC));
-    topics.push(Bytes32::from_u256_be(&b160_to_u256(caller)));
-    topics.push(message_hash);
-
-    system.io.emit_event(
-        // Use EVM to charge gas for this operation
-        ExecutionEnvironmentType::EVM,
-        resources,
-        &L1_MESSENGER_ADDRESS,
-        &topics,
-        // We are lucky that the encoding of the event is exactly same as encoding of the bytes in the calldata
-        &abi_encoded_message,
-    )?;
-
     Ok(Ok(message_hash))
```

Comment:
This breaks EVM compatibility - ideally we should not return anything here. 

As I understand, for that we'll have to calculate the hash twice (in L1Messenger contract and here, under the hood), since a caller can rely on the returned hash. Imo this is an acceptable overhead, otherwise any call to L1Messenger hook will create issues for tooling without custom integrations